### PR TITLE
4682 - Create Table Component

### PR DESCRIPTION
**Pull request by Miaplacidus**
The work here completes issue #4682. #4682 specifies the creation of a table component that can be used in the (not-yet-implemented) communities directory page. The table must provide the ability to set columns as sortable or not sortable. The table cells can display the following types of data: numbers, text, avatars alongside text, tags, and buttons. The table is initially sorted by its first column.

Note that, though our first use of this component will be the communities page, it can be used on any page, so long as data is passed to it in the manner specified in the comments in `/component_kit/cw_table.tsx`. 

## Link to Issue
Closes: #4682 

## Description of Changes
Creates a table with:
- the ability to configure a column as sortable
- the ability to configure a column as numerical (the text in numerical cells is displayed differently)
- the ability to display different types of data; specifically, text, numbers, avatars, tags, and buttons

## Test Plan
- navigate to [commonwealth demo components showcase](https://commonwealth-demo.herokuapp.com/components). The table component can be found at the bottom of the page
- Note that the table is initially sorted by the first column
- click on the sortable columns to check that the sorting occurs as expected
- click on the non-sortable columns (the ones without arrows) and check that nothing happens

### CODE REVIEW

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx:91]: is this line necessary? you dont use `_setData`, and `data` is passed to `useReactTable`, so I guess it could be simply rowData passed, and removing this line.

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx:84]: let's not use `React` prefix for useState, effect and other methods that can be imported. This just make the code more cluttered. Please change here and in other places in this PR

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx:165]: is this div wrapper necessary? Can't see `h-2` class anywhere defined in scss. Looks like leftover from some tailwind copied code

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx:176]: can you use clsx for dynamic classes?

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_table_data.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_table_data.tsx:109]: this file is created only for sake of showing table in component showcase, right? if yes, this should not be in component_kit and it should not be named cw_table_data. `commonwealth/packages/commonwealth/client/scripts/views/components/component_kit/helpers.ts` => this looks like a good place for that

File: packages/commonwealth/client/styles/components/component_kit/cw_table.scss
Comment (masvelio) [inline packages/commonwealth/client/styles/components/component_kit/cw_table.scss:14]: everywhere where you put html element tag prefix before the class eg. `div`, `span` - this is not necessary. Class selector is enough in this case. Please remove

File: packages/commonwealth/client/styles/components/component_kit/cw_table.scss
Comment (masvelio) [inline packages/commonwealth/client/styles/components/component_kit/cw_table.scss:56]: here and in other places below

**Review by masvelio — COMMENTED** [review]
Some small things that would be good to fix, but generally looks good!

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx:49]: Done!

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icons.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icons.tsx:79]: Linter set up properly and SVGs camel-cased!

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx:54]: I've created the CWTable directory, however, I've placed it in the component_kit directory. I created the new_designs directory to account for situations where we were creating new (storybook-displayed) components that weren't actually implemented (this was especially useful in cases where the name of the new component conflicted with the name of an existing component). This table component is very soon to be used as part of a new page. I should note that given that we're migrating away from Storybook, I've been thinking of what to do about the new_designs directory.

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx:165]: Deleted!

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx:176]: Can do!

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_table_data.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_table_data.tsx:109]: Good point. I've actually created a separate file called `showcase_helpers.tsx` since I don't want what are essentially helpers for displaying components on a developer/designer-used page mixed in with other helpers.

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icons.tsx
Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icons.tsx]: @Miaplacidus 
I still can see that prettier is not applied correctly eg in `CWTable`, [look here](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEcAeAHCAnGACSKAZ3wGEB1AFQEMAjAGzjwF48AKYAiegVwFsoASSgAzCABo82CAHcAItRjU8AXyR4aDOAAVpGIgEoWAPjzAAOlDxdi+ANpEcMAJZQA5pKJwYAZSeu3AF0WPB4vHyV4AB4-XACIxThjNjtLa2sLK3TrZwATdUheAWExOwAGQIA6AGs4AE9xNOy8XLgiMHURanovJusVJsCDSybCEi4i4hCwuABZOD4IKNJufig5OBEonihqqFkoSWooOuM7QOS+9iNmYyvrQrWSiEq+agw2NkKb00zm9Ow3h42CsbHuzT+-2a1DAYDajmwAGl6gVuDV6o0slD0gALODUVrYVH0Sp4glwbCY7HZOH0ejqNiuMQ-Mzg7FjfBgYGA2AAJVkISZL2kMkqOGcblc3Tw1CIeH58kU1AA3GyoRyZQA3JXYACq2HoIS52B5MAVlWo2qU2CIAH5KnZCui6oFVVA1f9nCJ2E6oPwKc4wEZIdTmoCYMDQR7qVFcs5NQR6LKiAA5ah8ODMADkfoz2EDWbuWNDUOAQsqbm8ADVujw4GwDANiyX0lEAPRxzVFlvZYbN6kqPBwHpML1sS06-X0AwhnvhyPsaPY2PxxPJtMZ7MT60AWlp9ELS5jAEErdRsKEDcxgNvz1PB0RnAAvTPAABMZUHbe7Pf+K4TYBJkQqbppmWbwGgMCFmWogvJWMA1rw9aNu2nY-r+1iofG6E9n2v6DsOXisv2obziCeDlvBiF1g2R79EeKhUiWYgQPARJ4BgehyrcHFcZUjwCJUeRMaGCB0IwsQuO4xKVAiShaCJUJNtieHNCoBiKeceCWH2ozQOM8mMNMXi8viYAwJojAcFcuRKuoIoKEoikCcQikkIk6iztYckBIpjFXNAkkBKQOLHJW6heL4-juIp8ErICCqzBArT0ngcU4HAiXJcODaxd4klwLkWUpeo8EFUVshJSluU2XAtA8G4llwOoMDYHWTHqSMWQanGRAYEmdSSYIhAhGwclyM4gLmc40ARa1AQsrOGrjZNcDTfpIReTK7TqP+a7ARuYGBtAe7QEorgUoWyzkMN0B0RRhCHVuJqyLqGAAEJIQArGUZRZvdx1QD4z5gUQbx0v9JHqkBIGblm54ijuPAYDuDB1pDLbflhXaKdYrQ7Xge2AeuoHZoDp2wNQF3YFdFC3e6UOeo9pPwy98gHJ9dY-X9AOEMDL7ZmD3QHvdxMHSzCOyDuuQHKjSEYz2WMdthuN4F0I67Z2+2w0dhAU+dUCXcY130-dOTM3DksyG9KZwDwrXdAALL9Cu-oD-Og+DIuM9DJOW2z0uy2jcCuyWStoX5XXZGRVgrVNLj6Q4TgTfHM1QIETRNk0MeLsWRMw09wArHwWCG7AlB1BgcCVE1Kg4ZhhlJGqUQwGSuT1xCjcVt4AAS+KEgA4tIyNEA2rzvGwbcUkPEDIyYucti3F61HU15T9gM-I0JuR1-dwDr5vGCkv3FJEOPHzr-PW2kUC5Fgj7y6t3gK9ryf2Db4OhQ+Bgxyv+S79fx-lAXeD9qT7zfkJIg2gkxwhxNwQkeBbR4D9HSPADIzZQn-Bg7EwBKh4OvhhGkBdQIFB6GgNgWZ176wQFBSQ69+KrEEnFY4kkGx4AAGTsLwFmY0CIdxYFcGxPAXhGDmR3PsQ2WYNLYOpNAUg9BAzVHUPQly3cLIQDcG4CS0U3A92OLkRg2AaqgJbCoZShDmgdwsYTPqxxtZPUoW-HcEEoJWOsWYEQjA0CmSgISe+7jqQqMYVABhkwNgiGPv-VWATrD0PSrAdAMBaImNwuYmJhM2y2KgEWGR-xwH-1CWsNRpAWFOGSek7ISC2C9X6tQQaTh6aTwgao+CggiDlTYbKYR813BGFtEglB05clQnUIMtJ7jsZuJbI2GR7ZW5TKhKpUxMzQFzOwAsgwKzlxtlbv3NxLdaDJVONGYAjcjxqOKjlJZ1JKgiiIOct4HwRRX3ujnfxGEl7P3qNeEUH8Fk4N+VRZwj4tCkGHD0MejyvjgpeSk-4bzhkN1yF81ewB9wf3sSzWySh9ZU0NjTf5PZgCePQD4vxiLmjotUS5cJ-FwXRPcei+JLjykVK2TEuZ7dhnXPwuy38azCXWB5UpPlzQ5mHNyMc-sczxJNzzsrLsTQ+wqBAOIEAEAMAJ2IMgUAVttDngQEQZAIBugyDqUatVtBsAwlqL4UCAAZC6yB1ZeEtdasAtrv4wgCMgVq7VUB8FoIVVouR7VhR4NQSsAAxHAbwYBSTcMa6g9sICqpADiGAfB6DkBxM4eAti4Q+ENXm+Mea6jGrAMBNNrgvC4F0JGt4zruiupAAAKyIGgYG7hGAAEUeCsTgE2kcaqf42gpMa2gdBhzOItSATigjyB5FbsgAAHGUEd0gvDkGtRgY1nE2gUk1IOtVABHft8BdAaqNSgWU4i4CFUKmmwEZ7Vr1rcI2pALq4Bqq8HwZwvq2rfpAI+btcBjzxvzPVC9FJHWGyHS2kDWi4B9oHfBoD8lF25GXUgN8aqHbOAUe4YuH7UBEC+mmmYmhr1frVZqOswhWiwB8GAfMmrjy+IiHURgaGzFAA)

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx]: In my opinion storybook is different story ;) Regardless we have it or not, we should distinct somehow new components from old ones. We have done it so far and changing this approach in half way is not ideal I guess.

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx]: not every now, but some of them still have it

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_table.tsx]: The CWTable directory has been moved to `/new_designs`.

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icons.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icons.tsx]: @masvelio Prettier has been applied in the most recent update.

**Comment by jessmart1213** [comment]
### Design review

**Sort**

- Update sort arrow on:hover and on:click zone to be limited to the Icon CTA. Currently the sorts can be activated or deactivated by clicking anywhere within the header cell.
    - I want to reserve clicking the header cell for more advanced table actions we will most likely implement in the future.

**Alignment**
- Update body cells to be aligned top. Currently, they are middle aligned.
    
    
image


**Body cell text**
- Ensure all body cells have text size of 14px ( Body 02 ).
- Update body cell text for cell type, User (no role) & User (role tag), to Body02/medium. Currently it is Body 01/reg or 16px height.
    
    image

---    

Questions: 

- Have tags had a review? Are these our new tags or old ones? Asking because the padding on this tag doesn’t match the padding within the figma designs.
    
    image
    
- I have the same question for buttons. I see our old buttons are being used in this table. Do they also need a design review & approval?

---

**Table body on hover and on click states**

- This is something that slipped through the cracks - where I designed a hover state for the Directory table but I don’t see it anywhere is the ticket. So I need to talk to product about next steps.
- See this [[loom](https://www.notion.so/Design-Solution-Related-communities-directory-page-add775d6bd6f425f9be51cd2bfb5e1cd?pvs=21)](https://www.notion.so/Design-Solution-Related-communities-directory-page-add775d6bd6f425f9be51cd2bfb5e1cd?pvs=21) at 5:00 min for table hover behavior.

**Comment by zakhap** [comment]
Is there a reason this is still in draft? Is it ready for Product review?

**Comment by Miaplacidus** [comment]
@jessmart1213 
1. the sorting mechanism now occurs only on clicking the column header icon (not on clicking any part of the header)
2. table body cells are top-aligned
3. font weights and sizing for body cells have been updated
4. I used older tags and buttons on the components page. However, it should be noted that either new or old buttons could have been used here. I have, however, updated the table on the components page to use the new tag and button designs. 
5. When table rows are hovered over, the background color changes. However, since there is no associated design, I had to guess at this color. It appeared to match that of the table header, so I went with that color (neutral 50)

**Comment by Miaplacidus** [comment]
@zakhap 
Was moved into draft to complete responses to reviews. Please do take a look at the table [here](https://commonwealth-demo.herokuapp.com/components)

**Review by masvelio — APPROVED** [review]
codewise looks good!

**Comment by ForestMars** [comment]
Thanks for the approval here Marcin. For the 2nd approver, Malik can you take a pass here, mainly to familiarize yourself with this component, **which will need to be updated for future Gating UI requirements** 

Details t/k, just want to make sure you've at least had a chance to review this.

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icon_lookup.ts
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icon_lookup.ts:40]: I am not sure if using colors in the lookup is a good idea. Curious for your thoughts @Miaplacidus @masvelio

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx:135]: I think we should add `columInfo` for memo dependency. 

If we want to only do this on render then maybe we can hold the prop in a reference `useRef`?

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx:147]: Maybe we should set debugTable: true only for dev mode. 

I was getting these in the console

image

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.scss
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.scss:4]: Can we make it fullwidth? On empty data it shows like this

image

**Review by mzparacha — COMMENTED** [review]
LGTM 🚀, some small changes requested.

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx
Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx:147]: agree

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx
Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx:135]: we have the code quality check that says that and we definitely should respect those warnings

React Hook useMemo has a missing dependency: 'columnInfo'. Either include it or remove the dependency array

File: packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icon_lookup.ts
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/cw_icons/cw_icon_lookup.ts:40]: Hey, @mzparacha!
I agree; I'd prefer to handle SVG colors using CSS. However, I believe I started doing this due to issues around updating the color/fill of certain SVGs via CSS. I'm not opposed to a conversation and/or second attempt at this (i.e. making a card to focus on consolidating SVGs and ensuring that they are colored as expected using CSS), but I think we should merge this work with the current icons.

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx:135]: `columnInfo` has been added to the dependency array.

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.tsx:147]: Hey @mzparacha @masvelio - `debugTable` only works in dev mode (if you take a look on the demo branch, you shouldn't see any console messages).

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.scss
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTable/CWTable.scss:4]: Good catch! Table is now fullwidth regardless of presence of data.

**Comment by Miaplacidus** [comment]
@mzparacha Thanks for your review!
I've resolved the issues you brought up above.

**Pull request merged**
